src/contexts/ui/LoadingOverlay.tsx

```diff
@@ -0,0 +1,112 @@
+import PropTypes, { InferProps } from 'prop-types';
+import React, {
+  createContext,
+  useCallback,
+  useContext,
+  useEffect,
+  useMemo,
+  useRef,
+  useState,
+} from 'react';
+
+import uniqid from 'uniqid';
+import { Backdrop, CircularProgress } from '@material-ui/core';
+
+import {
+  LoadingOverlayConfig,
+  LoadingOverlayControls,
+  LoadingOverlayCtx,
+  LoadingOverlayInitializer,
+} from '@app/types';
+
+const LoadingOverlayContext = createContext<LoadingOverlayCtx>(null);
+
+const LoadingOverlayProvider: React.FC = props => {
+  const [registeredLoaders, setRegisteredLoaders] = useState<LoadingOverlayConfig[]>([]);
+  const isOpen: boolean = registeredLoaders.some(loader => loader.active);
+
+  const setLoaderStatus = useCallback((key: string, value: boolean) => {
+    setRegisteredLoaders((loaders: LoadingOverlayConfig[]) => {
+      const dupLoaders = [...loaders];
+      const requestedLoader = dupLoaders.find((loader: LoadingOverlayConfig) => {
+        return loader.key === key;
+      });
+      if (requestedLoader) {
+        requestedLoader.active = value;
+      }
+      return dupLoaders;
+    });
+  }, []);
+
+  const deregisterLoader = useCallback((key: string) => {
+    setRegisteredLoaders((loaders: LoadingOverlayConfig[]) => {
+      const dupLoaders = [...loaders];
+      const requestedLoaderIndex: number = dupLoaders.findIndex((loader: LoadingOverlayConfig) => {
+        return loader.key === key;
+      });
+      if (requestedLoaderIndex >= 0) {
+        dupLoaders.splice(requestedLoaderIndex, 1);
+      }
+      return dupLoaders;
+    });
+  }, []);
+
+  const registerLoader = useCallback((): LoadingOverlayControls => {
+    const key: string = uniqid();
+    setRegisteredLoaders(prevState => [...prevState, { key, active: false }]);
+    return {
+      activate: (): void => setLoaderStatus(key, true),
+      deactivate: (): void => setLoaderStatus(key, false),
+      deregister: (): void => deregisterLoader(key),
+    };
+  }, [deregisterLoader, setLoaderStatus]);
+
+  const contextValue: LoadingOverlayInitializer = useMemo(
+    () => ({
+      register: registerLoader,
+    }),
+    [registerLoader],
+  );
+
+  return (
+    <LoadingOverlayContext.Provider value={contextValue}>
```

Comment:
no need to return an object on every context value

```suggestion
  return (
    <LoadingOverlayContext.Provider value={registerLoader}>
```

Reply:
I like returning objects for contexts because it makes it easier to think of a context as a "box" that has utilities in it for consumers to use. I know that there's just one thing being returned here but I'd like to keep with that pattern.

Reply:
I feel that may be over-generalizing how to approach contexts but honestly it's such a small difference that I could go either way - no need to change it.

As an aside, an instance I've found for a "single-value" context is an application might have a Logger context whose value changes due to configuration determined at runtime, but that context still only returned the logger object